TODO 15 (#9)
djuszczakk/tizen-docs: docs/application/web/guides/connectivity/bluetooth.md

      - write about setWriteValueRequestCallback() for descriptor
 -->
### Responding to read and write value requests from server clients

> 

Reviewer: Please, add a note (somewhere, not necessarily here), that the callbacks described here can also be registered by putting them in `BluetoothGATTServerCharacteristicInit` or `BluetoothGATTServerDescriptorInit`.

Reviewer: Please, grep for "desriptor" and replace all occurrences with "descriptor"

Author: done

> In order to react and respond to the read and write requsts from a client connected to the server running on the device, the callback functions needs to be set on the characteristics or on the descriptors.

Reviewer: ```suggestion
In order to react and respond to the read and write value requests from a client connected to the server running on the device, the callback functions need to be set on the characteristics or on the descriptors.
```

Author: done

> To set a callback function for read or write request on a characteristic or descriptor:

Reviewer: ```suggestion
To set a callback function for read or write value request on a characteristic or a descriptor:
```

Author: done

> 1. Prepare a GATT service containing at least one characteristic or a characteristic with at least one descriptor, that can be read or write:

Reviewer: ```suggestion
1. Prepare a GATT service containing at least one characteristic or a characteristic with at least one descriptor, that can be read or written:
```

Author: done

   ```
   var exampleDescriptor = {
      uuid: "0155",

>       /* Descriptor's permissions */

Reviewer: `readPermission` and `writePermission` are rather self-descriptive names. Please, consider removing this comment.

Author: done

      readPermission: true,
      writePermission: true
   };

   var exampleCharacteristic = {
      uuid: "0180",
      descriptors: [exampleDescriptor],

>       /* Characteristic's properties */

Reviewer: `isReadable` and `isWritable` are rather self-descriptive names. Please, consider removing this comment.

Author: done

      isReadable: true,
      isWritable: true,

>       /* Characteristic's permissions */

Reviewer: `readPermission` and `writePermission` are rather self-descriptive names. Please, consider removing this comment.

Author: done

      readPermission: true,
      writePermission: true
   };

   var gattService = {
      serviceUuid: "0955",

>       isPrimary: false,

Reviewer: Please, change this to `true` - according to the BT spec:

> A secondary service is a service that is only intended to be included from a primary service or another secondary service or other higher layer specification.

As there's no master service for gattService, it probably can't even be registered.

Author: OK. I'm changing this.
BTW:
I have successfully registered an only service which had `isPrimary` set to `false`.

>       services: [],

Reviewer: ```suggestion
      includedServices: [],
```

Author: done

      characteristics: [exampleCharacteristic]
   };
   ```

2. Register the service:
   ```
   var registerServiceSuccesCB = function() {
      console.log("Service successfully registered!");
   };

   var registerServiceErrorCB = function(e) {

>       console.log("Service not registered, error: " + e);

Reviewer: 1. Please, change `e` to something more descriptive (e.g. `error`).

2.  This call will log:

> Service not registered, error: [object Object]

which is not very helpful. Please, log error's message/name or both.

Author: done

   };

   var server = tizen.bluetooth.getGATTServer();

>    server.registerService(gattService, succesCB, errorCB);

Reviewer: ```suggestion
   server.registerService(gattService, registerServiceSuccesCB , registerServiceErrorCB);
```

Author: done

   ```

3. Start the GATT server:
   ```
   var serverStartSuccessCB = function() {
      console.log("Server started successfully!");
   };

   var serverStartErrorCB = function(e) {

>       console.error("Server didn't start, error: " + e);

Reviewer: This call will log:

    Server didn't start, error: [object Object]

which is not very helpful. Please, log error's message/name or both.

Author: done

   };

   server.start(serverStartSuccessCB, serverStartErrorCB);
   ```

> 4. To register the callback called when a client reads the value of the characteristic from the local GATT server, create the callback function and pass it as an argument to the `setReadValueRequestCallback()` method called on the characteristic object.

Reviewer: ```suggestion
4. To register the callback called when a client reads the value of the characteristic from the local GATT server, create the callback function and pass it as an argument to the `setReadValueRequestCallback()` method called on the characteristic object:
```

Author: done

   ```
   var characteristicReadRequestCallback = function(clientAddress, offset) {
      console.log(clientAddress + " requested to read characteristic's value with offset: " + offset);
      return new tizen.GATTRequestReply(0, "0x1234");
   };

>    server.services[0].characteristics[0].setReadValueRequestCallback(characteristicReadRequestCallback);

Reviewer: This function takes 5 arguments:

```
void setReadValueRequestCallback(ReadValueRequestCallback readValueRequestCallback, optional SuccessCallback? successCallback,
                                 optional ErrorCallback? errorCallback, optional SuccessCallback? sendResponseSuccessCallback,
                                 optional ErrorCallback? sendResponseErrorCallback);
```

Please, define all of them to show the guide readers how to use this function properly.

Author: I have added paragraph with all the callbacks. I made them common for all four functions in this guide section. Please review this change.

   ```

   > **Note**
   > A callback set with `setReadValueRequestCallback()` function overwrites any previously set `ReadValueRequestCallback` on this characteristic.

5. To register the callback called when a client writes a value of the characteristic of the local GATT server, create the callback function and pass it as an argument to the `setWriteValueRequestCallback()` method called on the characteristic object.
   ```
   var characteristicWriteRequestCallback = function(clientAddress, value, offset, replyRequired) {
      console.log(clientAddress + " requested to write characteristic's value: " + value +
                  " with offset: " + offset);

>       return replyRequired ? new tizen.GATTRequestReply(0, "0x1234") : null;

Reviewer: Adding the second argument in `GATTRequestReply` only makes sense for read value callbacks. Please, remove it.

Author: done

   };

>    server.services[0].characteristics[0].setWriteValueRequestCallback(characteristicWriteRequestCallback);

Reviewer: Same as before - please, define and add other callbacks.

Author: done

   ```

   > **Note**
   > A callback set with `setWriteValueRequestCallback()` function overwrites any previously set `WriteValueRequestCallback` on this characteristic.

> 6. To register the callback called when a client reads the value of the desriptor from the local GATT server, create the callback function and pass it as an argument to the `setReadValueRequestCallback()` method called on the desriptor object.

Reviewer 2: ```suggestion
6. To register the callback called when a client reads the value of the descriptor from the local GATT server, create the callback function and pass it as an argument to the `setReadValueRequestCallback()` method called on the descriptor object.
```

Author: done

   ```

>    var desriptorReadRequestCallback = function(clientAddress, offset) {

Reviewer: ```suggestion
   var descriptorReadRequestCallback = function(clientAddress, offset) {
```

Author: done

      console.log(clientAddress + " requested to read desriptor's value with offset: " + offset);
      return new tizen.GATTRequestReply(0, "0x1234");
   };

>    server.services[0].characteristics[0].descriptors[0].setReadValueRequestCallback(characteristicReadRequestCallback);

Reviewer: Same as before - please, define and add other callbacks.

Author: done

>    server.services[0].characteristics[0].descriptors[0].setReadValueRequestCallback(characteristicReadRequestCallback);

Reviewer: ```suggestion
   server.services[0].characteristics[0].descriptors[0].setReadValueRequestCallback(descriptorReadRequestCallback);
```

Author: done

   ```

   > **Note**

>    > A callback set with `setReadValueRequestCallback()` function overwrites any previously set `ReadValueRequestCallback` ot this descriptor.

Reviewer 2: ```suggestion
   > A callback set with `setReadValueRequestCallback()` function overwrites any previously set `ReadValueRequestCallback` on this descriptor.
```

Author: done

> 7. To register the callback called when a client writes a value of the desriptor of the local GATT server, create the callback function and pass it as an argument to the `setWriteValueRequestCallback()` method called on the desriptor object.

Reviewer 2: ```suggestion
7. To register the callback called when a client writes a value of the descriptor of the local GATT server, create the callback function and pass it as an argument to the `setWriteValueRequestCallback()` method called on the descriptor object.
```

Author: done

   ```
   var desriptorWriteRequestCallback = function(clientAddress, value, offset, replyRequired) {
      console.log(clientAddress + " requested to write desriptor's value: " + value +
                  " with offset: " + offset);

>       return replyRequired ? new tizen.GATTRequestReply(0, "0x1234") : null;

Reviewer: Adding the second argument in GATTRequestReply only makes sense for read value callbacks. Please, remove it.

Author: done

   };

>    server.services[0].characteristics[0].descriptors[0].setWriteValueRequestCallback(desriptorWriteRequestCallback);

Reviewer: Same as before - please, define and add other callbacks.

Author: done

   ```

   > **Note**

>    > A callback set with `setWriteValueRequestCallback()` function overwrites any previously set `WriteValueRequestCallback` on this desriptor.

Reviewer 2: ```suggestion
   > A callback set with `setWriteValueRequestCallback()` function overwrites any previously set `WriteValueRequestCallback` on this descriptor.
```

Author: done
